### PR TITLE
Allow students to delete student-logged grades for an assignment

### LINKED ISSUE #4237: Allow students to delete self-graded grades
### Expected Behavior
As a student, sometimes I submit a self-graded assignment and make a mistake - I need a way to fix it, by deleting my grade! Instructors should be able to declare at the Course level if students can un-delete or not. Students should not be able to delete after the deadline for the assignment.

### Actual Behavior
This feature is currently not built, requiring students to email their instructor for them to delete, so that students can then go re-grade.

### PR DESCRIPTION
**Pull request by adishy**
### Status
**READY**

### Description
* When a grade has been logged by a student for an assignment, it should be possible for them to delete the grade. Currently this is not possible and the instructor is required to delete the grade logged by the student if they want to resubmit the grade.
* Now a setting has been added under the "Advanced" tab in Course Settings that enables students to delete student logged grades for an assignment. A button that enables students to delete the grade they logged appears after they submit a grade. After they delete the grade they logged they can resubmit their grade as normal. They are allowed to repeat this as long as the assignment is open.

image

image

### Deploy Notes
* A boolean value called delete_student_logged_grade has been added to the Course model to track whether or not the instructors allow students to delete student logged grades.

### Migrations
YES

### Steps to Test or Reproduce
1. As an instructor, create an assignment and mark it as student logged. Ensure that the "Delete Student Logged Grades" setting  under the "Advanced" tab in Course Settings is checked.
2. As a student submit a grade for the assignment. Visit the submitted grade for the assignment (i.e. visit /assignments/\<id\>), you will see a button that says "Delete self-logged grade". Clicking the button will delete the grade that was submitted and will enable you to submit a grade for the assignment again.

### Impacted Areas in Application
* Course Model 
* Assignment Grades controller (i.e. controllers/assignments/grades_controller.rb)
* Course settings
* Viewing a grade as a student 

======================
Closes #4237

### CODE REVIEW

**Comment by maskib** [comment]
@adishy Sorry for the delay, I'm looking  into cases where the "delete grade" button disappears. Not sure what's causing it yet but I'll let you know!

**Comment by maskib** [comment]
@adishy It seems that the button to delete your grade does not display if an assignment both has 1. attached Learning Objectives and 2. Grade score levels. Can you let me know if you don't see the same thing?

**Comment by adishy** [comment]
@maskib 
Will look into it and report back!

**Comment by adishy** [comment]
@maskib @ebarroso7 
* While I couldn't reproduce the exact issue described above in the local environment, a related issue in the assignments controller (i.e. controllers/api/assignments_controller.rb) where linking learning objectives causes an error that prevents grade score levels from being created. This could have caused the assignment to not update and perhaps is the cause of the issue? Please do let me know if this solves the issue!

* The error caused by create_or_update_learning_objective_links in the assignments controller is described below:

```
=====================================

> undefined method `pluck' for nil:NilClass

app/controllers/api/assignments_controller.rb, line 106
-------------------------------------------------------

``` ruby
  101           learning_objective_linkable_id: @assignment.id,
  102           objective_id: objective_id,
  103           course_id: @assignment.course_id
  104       end unless assignment_params[:linked_objective_ids].nil?
  105   
> 106       @assignment.learning_objective_links.where.not(id: links.pluck(:id)).destroy_all if @assignment.learning_objective_links.any?
  107     end
  108   
  109     def assignment_params
  110       params.require(:assignment).permit(
  111         :accepts_attachments,

App backtrace
-------------

 - app/controllers/api/assignments_controller.rb:106:in `create_or_update_learning_objective_links'
 - app/controllers/api/assignments_controller.rb:46:in `update'
 - app/controllers/application_controller.rb:133:in `time_zone'
```

**Comment by maskib** [comment]
@adishy This issue looks to be resolved, thank you!

File: app/controllers/assignments/grades_controller.rb [hunk 1]
```diff
@@ -132,6 +132,33 @@ def self_log
...
+    @grade = Grade.find_by(assignment_id: @assignment.id, student_id: current_student.id)
+
+    if @grade.nil?
+      redirect_to assignments_path,
+        notice: "No self logged grade found."
+    end
+
+    @grade.destroy
+
+    ScoreRecalculatorJob.new(user_id: current_student.id,
```
Comment (ebarroso7) [inline app/controllers/assignments/grades_controller.rb]: Good catch to enqueue this now!

**Comment by ebarroso7** [comment]
Looks great! Nice migration too! @adishy

**Pull request merged**